csrc/multidevice/utils.cpp

```diff
@@ -399,27 +399,90 @@ bool haveDifferentShardings(
     }
 
     auto is_mapped_in_id_model =
-        [](IterDomain* a, IterDomain* b, const IdModel& id_model) -> bool {
-      if (a == nullptr && b == nullptr) {
+        [producer, consumer, mapped_p_logical_ids, mapped_c_root_ids](
+            IterDomain* p_loop_id,
+            IterDomain* c_loop_id,
+            const IdModel& id_model) -> bool {
+      if (p_loop_id == nullptr && c_loop_id == nullptr) {
         return true;
       }
 
-      if (a == nullptr || b == nullptr) {
+      if (p_loop_id == nullptr || c_loop_id == nullptr) {
         return false;
       }
 
+      if (consumer->definition()->isA<ViewOp>()) {
+        // Consider the reshape: [h]-> [a, h/a] where `h` and `a` are sharded,
+        // IdModel currently does not map the two DID axes. Similarly for the
+        // corresponding merge reshape. This is a temporary workaround to check
+        // if the reshape is resharding when sharding is on reshaped ids.
+
+        // Get the root iterdomains in the consumer that produce reshape logical
+        // Ids.
+        std::unordered_set<IterDomain*> c_reshape_ids;
+        for (auto id : consumer->getLogicalDomain()) {
+          if (id->isRFactorProduct() && id->definition() &&
+              !id->definition()->isA<Resize>()) {
+            auto root_ids = getInputsInTargetDomain(
+                id, {mapped_c_root_ids.begin(), mapped_c_root_ids.end()});
+            for (auto root_id : root_ids) {
+              c_reshape_ids.insert(root_id);
+            }
+          }
+        }
+
+        // Get the logical iterdomains in the producer that are reshaped.
+        const std::unordered_map<IterDomain*, IterDomain*>& c2p =
+            PairwiseLogicalDomainMap(producer, consumer)
+                .mapConsumerToProducer();
+        std::unordered_set<IterDomain*> p_reshape_ids;
+        for (auto id : c_reshape_ids) {
+          auto p_id = c2p.find(id);
+          if (p_id != c2p.end()) {
+            p_reshape_ids.insert(p_id->second);
+          }
+        }
+
+        auto p_transforms = DependencyCheck::getAllExprsBetween(
+            {p_reshape_ids.begin(), p_reshape_ids.end()}, {p_loop_id});
+        auto c_transforms = DependencyCheck::getAllExprsBetween(
+            {c_reshape_ids.begin(), c_reshape_ids.end()}, {c_loop_id});
+
+        if (p_transforms.size() > 0 && c_transforms.size() > 0) {
```

Comment:
@wujingyue I am not checking for these transforms, if we are always on the outer since it will be replaced soon. Does that work in the interim while your PR #3482 lands?

Reply:
SGTM!